### PR TITLE
[DeviceSanitizer] Support out-of-bounds on private memory

### PR DESCRIPTION
**Pull request by AllanZyne**
UR: https://github.com/oneapi-src/unified-runtime/pull/1676

Instrument "__asan_mem_to_shadow" to convert private address to its shadow memory address
Other steps are same with ASan on stack.

### CODE REVIEW

**Review by steffenlarsen — APPROVED** [review]
LGTM!

File: sycl/test-e2e/AddressSanitizer/out-of-bounds/private/multiple_private.cpp [hunk 1]
```diff
@@ -0,0 +1,76 @@
+// REQUIRES: linux, cpu
+// RUN: %{build} %device_asan_flags -Xarch_device "-mllvm=-asan-spir-privates=1" -DVAR=1 -O2 -g -o %t1
+// RUN: env SYCL_PREFER_UR=1 UR_LAYER_ASAN_OPTIONS=detect_privates:1 %{run} not %t1 2>&1 | FileCheck --check-prefixes CHECK,CHECK-VAR1 %s
```
Comment (AlexeySachkov) [inline sycl/test-e2e/AddressSanitizer/out-of-bounds/private/multiple_private.cpp:3]: I don't think that you need `SYCL_PREFER_UR` env variable. All existing PI plugins should be no-ops at this point, they simply redirect everything to UR.

**Review by AlexeySachkov — APPROVED** [review]
`llvm/` part LGTM, but I don't know details of asan implementation to do a proper review here. Does it make sense for us to create a separate GH team and assign it as a primary/secondary code owner for ASAN passes and tests?

File: sycl/test-e2e/AddressSanitizer/out-of-bounds/private/multiple_private.cpp (hunk 1 above)
Comment (kbenzie) [reply sycl/test-e2e/AddressSanitizer/out-of-bounds/private/multiple_private.cpp:3]: Until PI is fully replaced with UR this is required in order to actually use the UR loader, the sanitizer support is implemented as a UR layer which so is only loaded via the UR loader. It's necessary, for now, to actually test the sanitizer support.

Comment (AlexeySachkov) [reply sycl/test-e2e/AddressSanitizer/out-of-bounds/private/multiple_private.cpp:3]: > Until PI is fully replaced with UR this is required in order to actually use the UR loader, the sanitizer support is implemented as a UR layer which is only loaded via the UR loader. It's necessary, for now, to actually test the sanitizer support.

Oh, it is more complicated than I though. Got it, thanks for clarification

**Review by wenju-he — APPROVED** [review]
LGTM

**Comment by cdai2** [comment]
> `llvm/` part LGTM, but I don't know details of asan implementation to do a proper review here. Does it make sense for us to create a separate GH team and assign it as a primary/secondary code owner for ASAN passes and tests?

I support this proposal.  It will accelerate the development of Sanitizer related tasks.  
@bader what's your opinion ?

**Review by zhaomaosu — APPROVED** [review]
lgtm

**Comment by AllanZyne** [comment]
It seems the OCL RT in ["Linux", "gen12"] machine have not updated to latest version, that's why it failed.
Since only one machine, can we ignore this?

**Comment by yingcong-wu** [comment]
Could we add a return value check for the pfnGetSuggestedLocalWorkSize call and fallback to some alternative if fail to get result?

**Comment by AllanZyne** [comment]
> Could we add a return value check for the pfnGetSuggestedLocalWorkSize call and fallback to some alternative if fail to get result?

Good idea.

File: sycl/plugins/unified_runtime/CMakeLists.txt [hunk 2]
```diff
@@ -99,14 +99,14 @@ if(SYCL_PI_UR_USE_FETCH_CONTENT)
...
+  # Merge: 719bb9cd e2ffea69
   # Author: Kenneth Benzie (Benie) <k.benzie@codeplay.com>
-  # Date:   Fri Jun 21 10:22:52 2024 +0100
-  #     Merge pull request #805 from aarongreig/aaron/kernelSetArgIndirectionFix
-  #     Correct level of indirection used in KernelSetArgPointer calls.
-  set(UNIFIED_RUNTIME_TAG 1e9b1b493fe30e6236bf611ae6d82366c9376f6c)
+  # Date:   Thu May 23 10:53:03 2024 +0100
+  #     Merge pull request #1501 from RossBrunton/ross/kerneltests
+  #     [Testing] Spec clarifications and testing updates for kernel
+  set(UNIFIED_RUNTIME_TAG review/yang/private_memory)
```
Comment (callumfare) [inline sycl/plugins/unified_runtime/CMakeLists.txt:109]: https://github.com/oneapi-src/unified-runtime/pull/1676 has now merged, please resolve conflicts and update the UR tag:

```
  set(UNIFIED_RUNTIME_REPO "https://github.com/oneapi-src/unified-runtime.git")
  # commit 58ca3a34dea0f559b8d950bdfe7d5d8a610d3a94
  # Author: Yang Zhao <yang2.zhao@intel.com>
  # Date:   Thu Jun 27 20:26:17 2024 +0800
  #     [DeviceSanitizer] Support out-of-bounds on private memory (#1676)
  set(UNIFIED_RUNTIME_TAG 58ca3a34dea0f559b8d950bdfe7d5d8a610d3a94)
```

Comment (AllanZyne) [reply sycl/plugins/unified_runtime/CMakeLists.txt]: Done

Comment (callumfare) [reply sycl/plugins/unified_runtime/CMakeLists.txt]: Thanks!

**Review by callumfare — APPROVED** [review]

**Comment by callumfare** [comment]
@intel/llvm-gatekeepers Please merge

**Pull request merged**